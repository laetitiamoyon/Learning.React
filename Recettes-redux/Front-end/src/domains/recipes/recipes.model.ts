import { Ingredient } from '../ingredients/ingredients.model';

export interface RecipeIngredients extends Ingredient
{
  quantity? : number
}

export interface Recipe
{
  id : string,
  title : string,
  description : string,
  preparationTime? : string,
  cookingTime? : string,
  calories? : string,
  image? : string,
  ingredients : RecipeIngredients[]
}

export interface UpdateRecipeIngredientSuccessPayload
{
  recipeId : string,
  ingredient : Ingredient
}

export interface RemoveRecipeIngredientSuccessPayload
{
  recipeId : string,
  ingredientId : string
}