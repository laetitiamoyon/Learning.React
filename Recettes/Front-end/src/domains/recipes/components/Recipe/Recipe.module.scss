@import '../../../../styles/styles.scss'; 

.recipe
{
    @include widthHeight(240px, 248px);

    background-color: #e7f1f596;
    border-radius: $defaultBorderRadius;

    text-align: center;

    padding: $spacingMd $spacingXs;
    
    box-shadow: 0 1px 0 rgba(0,0,0,0.6);
}

.title
{
    cursor: pointer;

    @include text($textH5FontSize, undefined, $fontWeightSemiBold);
}
.title:hover { color: $secondaryColor;}

.image
{
    background-size: cover;
    @include widthHeight(150px, 150px);
}

.buttonContainer { @include flex(undefined, center); }

.viewDescriptionButton
{
    @include widthHeight(130px, 30px);
    @include borderAndRadius($defaultBorder, $defaultBorderRadius);
    @include text(undefined, $tertiaryColor, $fontWeightSemiBold);

    cursor: pointer;
    margin-left : 47px;
}
.viewDescriptionButton:hover{ @include buttonHover($secondaryColor, $defaultBorder, $primaryColor); }

.trashIcon
{
    @include superBackground(url(../images/icons/trash.png), cover);
    @include widthHeight(22px, 22px);
    margin-left : $spacingXs;

    cursor: pointer;
}