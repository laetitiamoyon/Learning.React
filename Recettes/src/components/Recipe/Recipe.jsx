import React from 'react'
import { useHistory } from "react-router-dom";
import styles from './Recipe.module.css'

const Recipe = ({id, title, image}) =>
{
    const history = useHistory();

    const onClick = () => history.push(`/description-de-la-recette/${id}`, { id : id })

    return <div className={styles.recipeContainer}>
        <div className={styles.recipeElement}>
            <div className={styles.recipeImage} src={image}></div>
            <div className={styles.recipeTitle}>{title}</div>
            <div className={styles.recipeButton} onClick={onClick}>Voir la recette</div>
        </div>
    </div>
}

export default Recipe;