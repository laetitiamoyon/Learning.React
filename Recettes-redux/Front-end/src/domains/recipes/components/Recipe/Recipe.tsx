import { FC } from 'react';
import { useHistory } from 'react-router-dom';
import { Recipe as RecipeModel} from '../../recipes.model';
import styles from './Recipe.module.scss'
import { useDispatch } from 'react-redux';
import { removeRecipeRequestAction, removeRecipeIngredientRequestAction } from '../../recipes.actions';
import { Highlight } from '../../../../shared/utils/Highlight';

interface Props extends RecipeModel
{
  searchTerm : string,
  color: string
}

const Recipe : FC<Props> = ({ id, title, image, searchTerm, color }) => 
{
  const history = useHistory()

  const onClickToTheDescription = () => history.push(`/description-de-la-recette/${id}`, { id : id })

  const recipeDispatch = useDispatch()
  const ingredientDispatch = useDispatch()

  const removeRecipe = () => 
  {
    recipeDispatch(removeRecipeRequestAction(id)) 
    ingredientDispatch(removeRecipeIngredientRequestAction(id))
  }

  return <div className={styles.container}>
      <div className={styles.recipe}>
        <Highlight searchTerm={searchTerm} text={title} color={color}/>
        <img className={styles.image} alt="" src={image} /> 
        <div className={styles.buttonContainer}>
          <button className={styles.viewDescriptionButton} onClick={onClickToTheDescription}>Voir la recette</button>
          <div className={styles.trashIcon} onClick={removeRecipe}></div>
        </div>
      </div>
    </div>
};

export default Recipe;