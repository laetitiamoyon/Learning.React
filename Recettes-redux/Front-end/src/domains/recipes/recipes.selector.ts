import { ApplicationState } from '../root/root.state';
import { RecipeState } from './recipes.state';

export const selectRecipes = (state : ApplicationState) : RecipeState  => state.recipes

export const selectInformationMessage = (state : ApplicationState) : string | undefined => state.recipes.informationMessage