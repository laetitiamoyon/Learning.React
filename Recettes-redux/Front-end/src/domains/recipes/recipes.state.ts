import { ActionStatus, IActionMetadata } from "../../shared/domains/Redux/redux.model";
import { Recipe } from "./recipes.model";

export interface RecipeState extends IActionMetadata
{ 
    recipes : Recipe[]
    informationMessage? : string
};

export const initialRecipes : RecipeState = 
{
    status : ActionStatus.Loading,
    error : undefined,
    recipes : [],
    informationMessage : undefined
}
