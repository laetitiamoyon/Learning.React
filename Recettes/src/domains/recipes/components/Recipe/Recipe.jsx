import React, { useContext } from 'react'
import { useHistory } from "react-router-dom";
import { RecipeContext } from '../../recipes.provider';
import styles from './Recipe.module.css'

const Recipe = ({id, title, image}) =>
{
    const history = useHistory();

    const onClick = () => history.push(`/description-de-la-recette/${id}`, { id : id })

    const { dispatch } = useContext(RecipeContext)
    const removeRecipe = () => dispatch({ type: 'REMOVE_RECIPE', payload : { id } })


    return <div className={styles.recipeContainer}>
        <div className={styles.recipeElement}>
            <div className={styles.recipeImage} src={image}></div>
            <div className={styles.recipeTitle}>{title}</div>
            <div className={styles.recipeButton} onClick={onClick}>Voir la recette</div>
            <div className={styles.removeButton} onClick={removeRecipe}>Supprimer</div>
        </div>
    </div>
}

export default Recipe;