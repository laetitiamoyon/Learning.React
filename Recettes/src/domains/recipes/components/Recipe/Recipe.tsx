import { FC, useContext } from 'react'
import { useHistory } from "react-router-dom";
import { removeRecipeAction } from '../../recipes.action';
import { RecipeContext } from '../../recipes.context';
import { Recipe as RecipeModel } from '../../recipes.model'
import styles from './Recipe.module.scss'

const Recipe : FC<RecipeModel> = ({id, title, imagePath, imageData}) =>
{
    const history = useHistory();
    const seeRecipeOnClick = () => history.push(`/description-de-la-recette/${id}`, { id : id })
    const removeRecipe = () => dispatch(removeRecipeAction(id))
    const { dispatch } = useContext(RecipeContext)

    return <div className={styles.recipeContainer}>
        <div className={styles.removeButton} onClick={removeRecipe}></div>
       
        <div className={styles.recipeElement}>
            <img className={styles.recipeImage} alt='' src={imageData} style={{ backgroundImage : `url(${imagePath})`}} />
            <div className={styles.recipeTitle}>{title}</div>
            <div className={styles.viewRecipeButton} onClick={seeRecipeOnClick}>Voir la recette</div>
        </div>
    </div>
}

export default Recipe;