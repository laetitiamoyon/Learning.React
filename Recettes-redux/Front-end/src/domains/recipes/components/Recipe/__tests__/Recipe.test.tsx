import {useDispatch} from "react-redux";
import {fireEvent, render} from "@testing-library/react";
import Recipe from "../../../../recipes/components/Recipe/Recipe";
import {
    removeRecipeIngredientRequestAction,
    removeRecipeRequestAction,
} from "../../../recipes.actions";
import {recipeMock} from "../../../recipes.mock";
import {useHistory} from "react-router-dom";

jest.mock('react-redux')
const mockUseDispatch = useDispatch as jest.MockedFunction<typeof useDispatch>
const mockDispatch = jest.fn()

const history = useHistory()
const mockOnClickToTheDescription = jest.fn(history.push)
jest.mock('react-router-dom', () => ({

    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockOnClickToTheDescription,
    }),
}));

describe("Recipe", () => {

    beforeEach(() => (mockUseDispatch as jest.Mock).mockReturnValue(mockDispatch))

    it("should match snapshot", () => {
        // Given & When
        const {container} = render(<Recipe {...recipeMock} color={""} searchTerm={""}/>)

        // Then
        expect(container).toMatchSnapshot()
    })

    it("should dispatch removeRecipeRequestAction and removeRecipeIngredientRequestAction when trash icon is clicked ", () => {
        // Given
        const {container} = render(<Recipe {...recipeMock} color={""} searchTerm={""}/>)
        const trashIcon = container.getElementsByTagName('div')[3]

        // When
        fireEvent.click(trashIcon)
        // Then
        expect(mockDispatch).toBeCalledWith(removeRecipeRequestAction(recipeMock.id))
        expect(mockDispatch).toBeCalledWith(removeRecipeIngredientRequestAction(recipeMock.id));
    })

    it("should redirected to description of the recipe when viewDescription Button is clicked ", () => {
        // Given
        const { getByText } = render(<Recipe {...recipeMock} color={""} searchTerm={""}/>)
        const button = getByText("Voir la recette")


        // When
        fireEvent.click(button)

        // Then
        expect(mockOnClickToTheDescription).toHaveBeenCalledWith('`/modification-de-la-recette/${id}`, { id : id }');
    })
})