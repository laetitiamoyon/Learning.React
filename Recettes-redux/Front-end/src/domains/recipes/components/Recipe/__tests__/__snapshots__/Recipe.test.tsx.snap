// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe should match snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="recipe"
    >
      <div>
        <span
          style="color :"
        />
        P
        <span
          style="color :"
        />
        i
        <span
          style="color :"
        />
        z
        <span
          style="color :"
        />
        z
        <span
          style="color :"
        />
        a
        <span
          style="color :"
        />
         
        <span
          style="color :"
        />
        M
        <span
          style="color :"
        />
        a
        <span
          style="color :"
        />
        r
        <span
          style="color :"
        />
        g
        <span
          style="color :"
        />
        h
        <span
          style="color :"
        />
        e
        <span
          style="color :"
        />
        r
        <span
          style="color :"
        />
        i
        <span
          style="color :"
        />
        t
        <span
          style="color :"
        />
        a
        <span
          style="color :"
        />
      </div>
      <img
        alt=""
        class="image"
        src="../images/pizza.jpg"
      />
      <div
        class="buttonContainer"
      >
        <button
          class="viewDescriptionButton"
        >
          Voir la recette
        </button>
        <div
          class="trashIcon"
        />
      </div>
    </div>
  </div>
</div>
`;
